### DISCUSSION TITLE
How to get the raw expand snippet text?

### DISCUSSION THREAD

**Discussion by xzbdmw** [L3MON4D3/LuaSnip#1146] [General]
Say I choose `local` in nvim cmp, after expansion, it will insert
```lua 
local function()

end
```
<img width="261" alt="截屏2024-03-24 20 20 25" src="https://github.com/L3MON4D3/LuaSnip/assets/97848247/7ad9e82a-61d6-4418-b461-43d038e44fce">

how do I get the raw text so that I can make ghost text display `function()` instead of `function$1($2)`, I want to change their implementation to expand early

**Comment by xzbdmw** [comment]
I manually concat static_text in https://github.com/L3MON4D3/LuaSnip/blob/a7a4b4682c4b3e2ba82b82a4e6e5f5a0e79dec32/lua/luasnip/nodes/node.lua#L65
using this function
```lua
local function put(lines, text, pos)
	local lineIndex = pos[1]
	local colIndex = pos[2] + 1
	for i, txt in ipairs(text) do
		if lines[lineIndex] then
			lines[lineIndex] = lines[lineIndex] .. txt
		else
			lines[lineIndex] = txt
		end

		if i < #text then
			lineIndex = lineIndex + 1
			colIndex = 1
		else
			colIndex = colIndex + #txt
		end
	end
end

```
can achieve https://github.com/hrsh7th/nvim-cmp/issues/1862#issuecomment-2016838904
It there any similar api I can use or can you add something similar to api?

**Reply by L3MON4D3** [reply]
Hey, you can try `snip:get_static_text()`, that should work if you call it on a snippet obtained from `ls.get_snippets`

**Reply by L3MON4D3** [reply]
Very cool idea btw :+1:

**Reply by xzbdmw** [reply]
Hi, get_static_text don't have indentation information about current cursor position,
```txt
get_ghost_text text: { "r index, value in ipairs(t) do", "\t", "end" }
get_ghost_text res: {
  [451] = "r index, value in ipairs(t) do",
  [452] = "\t\t",
  [453] = "\tend"
}
```
When I type `for` after a tab, `text` is `get_static_text`'s return value and `res` the return value of `put` function I comment above which manually concat spaces.

**Reply by L3MON4D3** [reply]
Ahhh, right.. I think you might be able to fix this by calling `get_static_text` on `snip:copy():indent("<indentation at cursor>")`  instead of `snip` (copy so the indentation does not end up in the stored snippet)

**Reply by xzbdmw** [reply]
Thanks, that work nice!
